pkg/vgmanager/vgmanager_controller.go

```diff
@@ -0,0 +1,206 @@
+	}
+
+	// load lvmd config
+	lvmdConfig := &lvmdCMD.Config{
+		SocketName: topolvm.DefaultLVMdSocket,
+	}
+
+	cfgBytes, err := os.ReadFile(ConfigFilePath)
+	if os.IsNotExist(err) {
+		r.Log.Info("lvmd config file doesn't exist, will create")
+		err = yaml.Unmarshal(cfgBytes, &lvmdConfig)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	} else if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	oldLVMDConfig := *lvmdConfig
+
+	// avoid having to iterate through device classes multiple times, map from name to config index
+	deviceClassMap := make(map[string]int)
+	for i, deviceClass := range lvmdConfig.DeviceClasses {
+		deviceClassMap[deviceClass.Name] = i
+	}
+
+	// filter out block devices
+	remainingValidDevices, delayedDevices, err := r.filterAvailableDevices(blockDevices)
+	if err != nil {
+		_ = err
+	}
+
+	for _, deviceClass := range lvmCluster.Spec.DeviceClasses {
+		// ignore deviceClasses whose LabelSelector doesn't match this node
+		// NodeSelectorTerms.MatchExpressions are ORed
+		node := &corev1.Node{}
+		selectsNode, err := NodeSelectorMatchesNodeLabels(node, deviceClass.NodeSelector)
+		if err != nil {
+			r.Log.Error(err, "failed to match nodeSelector to node labels")
+			continue
+		}
+		if !(selectsNode && ToleratesTaints(deviceClass.Tolerations, node.Spec.Taints)) {
+			continue
+		}
+		_, found := deviceClassMap[deviceClass.Name]
+		if !found {
+			lvmdConfig.DeviceClasses = append(lvmdConfig.DeviceClasses, &lvmd.DeviceClass{
+				Name:        deviceClass.Name,
+				VolumeGroup: deviceClass.Name,
+			})
+		}
+		var matchingDevices []internal.BlockDevice
+		remainingValidDevices, matchingDevices, err = filterMatchingDevices(remainingValidDevices, deviceClass)
+		if err != nil {
+			r.Log.Error(err, "could not filterMatchingDevices")
+			continue
+		}
+		if len(matchingDevices) > 0 {
+			// create/update VG and update lvmd config
+			err = r.addMatchingDevicesToVG(matchingDevices, deviceClass.Name)
+			if err != nil {
+				r.Log.Error(err, "could not prepare volume groups")
+				continue
+			}
+		}
+	}
+
+	// apply and save lvmconfig
+	// pass config to configChannel only if config has changed
+	if !cmp.Equal(oldLVMDConfig, lvmdConfig) {
+		r.Log.Info("updating lvmd config")
+		out, err := yaml.Marshal(lvmdConfig)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		err = os.WriteFile(ConfigFilePath, out, 0600)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// requeue faster if some devices are too recently observed to consume
+	requeueAfter := time.Minute * 2
+	if len(delayedDevices) > 0 {
+		requeueAfter = time.Second * 30
+	}
+	return ctrl.Result{RequeueAfter: requeueAfter}, nil
+}
+
+// filterMatchingDevices returns unmatched and matched blockdevices
+func filterMatchingDevices(blockDevices []internal.BlockDevice, lvmCluster lvmv1alpha1.DeviceClass) ([]internal.BlockDevice, []internal.BlockDevice, error) {
+	// currently just match all devices
+	return []internal.BlockDevice{}, blockDevices, nil
```

Comment:
Does this skip the installation disk?

Reply:
No. Right now there is no filter. we just return all the devices back.

Reply:
I spoke too soon. Yes there is a check.  There is a filter called  `noBiosBootInPartLabel`. It checks for `bios` or `boot` in the disk part label. Based on our experience with LSO so far, it might not cover all the cases.